src/api.jl

```diff
@@ -0,0 +1,24 @@
+module api
+
+paths = ByteString[]
+const libvulkan = Libdl.find_library(["libvulkan", "vulkan", "libvulkan.so.1"], paths)
+@assert libvulkan != ""
+
+function _vk_func(func, ret_type, arg_types)
+  local args_in = Symbol[symbol("arg$i::$T")
+  for (i, T) in enumerate(arg_types.args)]
+  esc(quote
+    function $func($(args_in...))
+      ccall(($(string(func)), libvulkan),
+      $ret_type,
+      $arg_types,
+      $(args_in...))
+    end
+  end)
+end
```

Comment:
> From providing a library perspective I would like to have the method definition to be as tight as possible, to prevent errors.

The issue is that you will only introduce more error this way. As an example,

`ccall(:cfunction, Void, (Ptr{Void},), wrapper.ptr)` is invalid if the `.ptr` is managed by a finalizer on `wrapper`. Julia is free to GC `wrapper` (and therefore `wrapper.ptr`) before the `ccall` finishes (e.g. if convertion of other arguments allocates, or if the `cfunction` calls back to julia runtime, or if a GC is triggerred in a different thread for threading build).

> So I take it I need to add cconvert and unsafe_convert for all kinds of types we might pass as references? I thought as long as all types involved are isbits Julia should be able to do the transformation automatically.

Not for the case I mentioned above. One of the correct way to deal with it is to define `cconvert(::Type{Ptr{Void}}, wrapper::WrapperType) = wrapper` and `unsafe_convert(::Type{Ptr{Void}}, wrapper::WrapperType) = wrapper.ptr`. See my [LibArchive.jl](https://github.com/yuyichao/LibArchive.jl/blob/4112772e8d1d7124c896bf98c5baee13149d3f8d/src/reader.jl#L52) for an example.